### DISCUSSION TITLE
How to Configure ReRAM as a Processing Unit

### DISCUSSION THREAD

**Discussion by 24367452** [IBM/aihwkit#686] [Q&A] [answered]
I want to use ReRAM as a processing unit for simulating CNNs, but I don't know what interface to use to configure it.
At first, I thought that configuring the noise model of the inference RPU also configured the processing unit as follows:
```
    my_rpu_config = InferenceRPUConfig(noise_model=ReRamWan2022NoiseModel())
    my_rpu_config.mapping.digital_bias = True
    my_rpu_config.mapping.out_scaling_columnwise = True
    my_rpu_config.mapping.learn_out_scaling = True
    my_rpu_config.mapping.weight_scaling_omega = 1.0
    my_rpu_config.mapping.weight_scaling_columnwise = False
    my_rpu_config.mapping.max_input_size = 512
    my_rpu_config.mapping.max_output_size = 512

    my_rpu_config.noise_model = ReRamWan2022NoiseModel()
    my_rpu_config.remap.type = WeightRemapType.CHANNELWISE_SYMMETRIC
    my_rpu_config.clip.type = WeightClipType.LAYER_GAUSSIAN
    my_rpu_config.clip.sigma = 2.5

    # train input clipping
    my_rpu_config.forward.noise_management = NoiseManagementType.NONE
    my_rpu_config.forward.bound_management = BoundManagementType.NONE
    my_rpu_config.forward.out_bound = 10.0  # quite restrictive
    my_rpu_config.pre_post.input_range.enable = True
    my_rpu_config.pre_post.input_range.manage_output_clipping = True
    my_rpu_config.pre_post.input_range.decay = 0.001
    my_rpu_config.pre_post.input_range.input_min_percentage = 0.95
    my_rpu_config.pre_post.input_range.output_min_percentage = 0.95

    my_rpu_config.modifier.type = WeightModifierType.ADD_NORMAL
    my_rpu_config.modifier.std_dev = 0.1

    my_rpu_config.runtime.data_type = DATA_TYPE
```
But then I realised that it seems that using

```rpu_config = SingleRPUConfig(device=ReRamESPresetDevice())```

It is the way to configure ReRAM as a processing unit.

 In the 06 example (Hardware-aware training of a simulated CNN - [examples/06_lenet5_hardware_aware.py](https://github.com/IBM/aihwkit/blob/master/examples/06_In [examples/06_lenet5_hardware_aware.py](_lenet5_hardware_aware.py)) SingleRPUConfig is not used, so how is the processing unit configured in this example?Does it use ReRAM or PCM?

If you know the right way to configure the ReRAM processing unit in a CNN, please let me know, thank you very much!

**Comment by 24367452** [comment]
06 Example (Hardware-aware training of simulated CNNs -https://github.com/IBM/aihwkit/blob/master/examples/06_lenet5_hardware_aware.py#L79)

**Comment by maljoras** [comment]
Hi @24367452 , what are you trying to do exactly, in-memory inference or in-situ training? 

We have and ReRAM model that is used during in-situ training (that is in-memeory training that uses incremental updates on ReRAM during backpropgation on the analog crossbar arrays) and an ReRAM model that is used for inference only.
 In the inference case, training is done with exact weight update (that is, in software) and the converged software is then only transferred (programmed) to the conductances in the crossbar array. Then inference is simulated using the noisy analog chip. This we refer to as "inference hardware".  In the latter case you need to use the `ReRAMWan2022NoiseModel` together with the `InferenceRPUConfig` as you did. You need to use `model.drift_analog_weights()` after the training to simulate the noise and programming of the conductances. After that you can test the accuracy. This is show-cased in the Example 6

**Comment by maljoras** [comment]
Please also refer to the tutorial paper, where it is explained in more detail, see [here](https://pubs.aip.org/aip/aml/article/1/4/041102/2923573/Using-the-IBM-analog-in-memory-hardware)

**Comment by 24367452** [comment, marked as answer]
Hello @maljoras ,Thank you for your interest, I would like to use the ReRAM model for in-situ training and evaluate the impact of non-ideal factors on the accuracy of the model.Does the following code fulfil my requirements?
```
    dmodel = SimpleModel()
    dmodel.to(device)
    criterion = nn.MSELoss()
    optimizer = create_digital_optimizer(dmodel, 0.01)
    algorithm='ttv2'
    rpu_config = build_config(algorithm,device=ReRamArrayHfO2PresetDevice)
    amodel = convert_to_analog(dmodel, rpu_config)
```

**Reply by maljoras-sony** [reply]
That looks reasonable!

**Comment by PabloCarmona** [comment]
Hello @24367452! The inputs from @maljoras-sony help you out? Need any more help or we can consider this closed? Thanks for taking your time using and testing the project!
